static/js/browser2/observation_chart.tsx

```diff
@@ -157,29 +155,34 @@ export class ObservationChart extends React.Component<
       .attr("style", "position: absolute; display: none; z-index: 1");
   }
 
-  private handleDotHover = (svgContainerId: string) => (
-    dotData: DotDataPoint
+  private handleDotHighlight = (svgContainerId: string) => (
+    dotData: DataPoint,
+    datapointX: number,
+    datapointY: number,
+    leftMargin: number,
+    bottomMargin: number
   ): void => {
+    const text = `${dotData.label}: ${dotData.value} ${this.getUnits()}`;
     const tooltipSelect = d3
       .select("#" + svgContainerId)
-      .select(`#${TOOLTIP_ID}`);
-    const text = `${dotData.label}: ${dotData.value}${this.getUnits()}`;
-    const tooltipHeight = (tooltipSelect.node() as HTMLDivElement).clientHeight;
-    const offset = 10;
-    const leftOffset = offset;
-    const topOffset = -tooltipHeight - offset;
-    tooltipSelect
-      .text(text)
-      .style("left", d3.event.offsetX + leftOffset + "px")
-      .style("top", d3.event.offsetY + topOffset + "px")
-      .style("display", "block");
-    this.updateErrorMessage("");
-  };
-
-  private handleDotLeave = (svgContainerId: string) => (): void => {
-    d3.select("#" + svgContainerId)
       .select(`#${TOOLTIP_ID}`)
-      .style("display", "none");
+      .text(text);
+    const rect = (tooltipSelect.node() as HTMLDivElement).getBoundingClientRect();
+    const width = rect.width;
+    const height = rect.height;
+    const topOffset = 5;
+    let left = datapointX - width / 2;
```

Comment:
some comments here would be helpful to explain the expected outcome.

Reply:
done